k8s/src/main/scala/io/buoyant/k8s/ServiceNamer.scala

```diff
@@ -0,0 +1,174 @@
+    override protected def mkResource(name: String): NsListResource[Service, ServiceWatch, ServiceList] =
+      mkApi(name).services
+
+    override protected def mkCache(name: String): ServiceCache =
+      new ServiceCache(name)
+
+    override protected def update(cache: ServiceCache)(event: ServiceWatch): Unit =
+      cache.update(event)
+
+    override protected def initialize(
+      cache: ServiceCache,
+      list: ServiceList
+    ): Unit = cache.initialize(list)
+  }
+}
+
+class ServiceCache(namespace: String) {
+
+  def get(serviceName: String, portName: String): Var[Option[Var[Address]]] = synchronized {
+    // we call this unstable because every change to the Address will cause
+    // the entire Var[Option[Address]] to update.
+    val unstable = cache.get(serviceName) match {
+      case Some(ports) => ports.map(_.get(portName))
+      case None =>
+        val ports = Var(Map.empty[String, Address])
+        cache += serviceName -> ports
+        ports.map(_.get(portName))
+    }
+
+    // We can stabilize this by changing the type to Var[Option[Var[Address]]].
+    // If this service port is created or deleted, the outer Var will update.
+    // If the address of the service port changes, only the inner Var will
+    // update.
+    val init = unstable.sample().map(Var(_))
+    Var.async[Option[VarUp[Address]]](init) { update =>
+      // the current inner Var, null if the outer Var is None
+      @volatile var addr: VarUp[Address] = null
+
+      unstable.changes.respond {
+        case Some(address) if addr == null =>
+          // Address created
+          addr = Var(address)
+          update() = Some(addr)
+        case Some(address) =>
+          // Address modified
+          addr() = address
+        case None =>
+          // Address deleted
+          addr = null
+          update() = None
+      }
+    }
+  }
+
+  private[this]type VarUp[T] = Var[T] with Updatable[T]
+
+  private[this] var cache = Map.empty[String, VarUp[Map[String, Address]]]
+
+  private[this] def extractPorts(service: Service): Map[String, Address] = {
+    val ports = for {
+      meta <- service.metadata.toSeq
+      name <- meta.name.toSeq
+      status <- service.status.toSeq
+      lb <- status.loadBalancer.toSeq
+      ingress <- lb.ingress.toSeq.flatten
+      spec <- service.spec.toSeq
+      port <- spec.ports
+    } yield port.name -> Address(new InetSocketAddress(ingress.ip, port.port))
+    ports.toMap
+  }
+
+  def initialize(list: ServiceList): Unit = synchronized {
+    val services = for {
+      service <- list.items
+      meta <- service.metadata
+      name <- meta.name
+    } yield name -> Var(extractPorts(service))
+    cache = services.toMap
+  }
+
+  def update(watch: ServiceWatch): Unit = synchronized {
+    watch match {
+      case ServiceAdded(service) =>
+        for {
+          meta <- service.metadata
+          name <- meta.name
+        } {
+          log.info("k8s added service: %s", name)
```

Comment:
i imagine this logging (and elsewhere) would be more useful if it included the namespace?